.github/workflows/test.yml

```diff
@@ -0,0 +1,78 @@
+# Copyright      2021  Fangjun Kuang (csukuangfj@gmail.com)
+
+# See ../../LICENSE for clarification regarding multiple authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+name: test
+
+on:
+  push:
+    branches:
+      - master
+  pull_request:
+    branches:
+      - master
+
+jobs:
+  test:
+    runs-on: ${{ matrix.os }}
+    strategy:
+      matrix:
+        os: [ubuntu-18.04, macos-10.15]
+        python-version: [3.6, 3.7, 3.8, 3.9]
+        torch: ["1.8.1"]
+        k2-version: ["1.2.dev20210723"]
+      fail-fast: false
+
+    steps:
+      - uses: actions/checkout@v2
+        with:
+          fetch-depth: 0
+
+      - name: Setup Python ${{ matrix.python-version }}
+        uses: actions/setup-python@v1
+        with:
+          python-version: ${{ matrix.python-version }}
+
+      - name: Install Python dependencies
+        run: |
+          python3 -m pip install --upgrade pip pytest kaldialign
+          pip install k2==${{ matrix.k2-version }}+cpu.torch${{ matrix.torch }} -f https://k2-fsa.org/nightly/
+
+          # Don't use: pip install lhotse
+          # since it installs a version of PyTorch that is not predictable
```

Comment:
Any suggestions how to fix it in on Lhotse side? I ran into this issue too..

Reply:
The following are the installation logs from GitHub actions:
```
Requirement already satisfied: torch>=1.7.1 in /opt/hostedtoolcache/Python/3.7.11/x64/lib/python3.7/site-packages (from lhotse===0.8.0.dev-e95d134-clean) (1.8.1)
Collecting torchaudio>=0.7.1
  Downloading torchaudio-0.9.0-cp37-cp37m-manylinux1_x86_64.whl (1.9 MB)

...

Collecting torch>=1.7.1
  Downloading torch-1.9.0-cp37-cp37m-manylinux1_x86_64.whl (831.4 MB)
```

---

I think the line `torchaudio>=0.7.1` in lhotse's requirements.txt is problematic.

Imagine the following case:
(1) A user has not installed torchaudio before
(2) The user has installed torch==1.8.1
(3) If he tries to install lhotse, pip sees that torch 1.8.1 is newer than 1.7.1, so torch>=1.7.1 is skipped
(4) pip sees `torchaudio>=0.7.1`, it tries to install the latest torchaudio, i.e., 0.9.0
(5) torchaudio 0.9.0 depends on torch 1.9.0
(6) pip replaces the already installed torch 1.8.1 with 1.9.0. OOPS

----

One solution I would recommend is to modify lhotse's setup.py:
(1) Detect if PyTorch is installed. If it is not installed or its version is < 1.7.1, then no changes to lhotse are needed
(2) If it is installed and its version is >= 1.7.1, set the version of torchaudio to match the version of PyTorch.
     For example, if torch==1.8.1 is detected, fix the version of torchaudio to 0.8.0.

Reply:
Thanks, good idea. Will do.

Reply:
... I think a better solution might be to remove torch and torchaudio from requirements.txt. 

I started implementing what you suggested but I hate the idea of keeping track of which version of pytorch corresponds to which version of torchaudio. It would mean every time they release even a minor update, I'd have to update Lhotse's setup.py. 

If this causes issues I'd rather leave it up to the user to install the pytorch and torchaudio version they want, and if they forget to do it, run into an import error in Lhotse (at least it's fairly obvious how to fix that one).

BTW at some point I might try to decouple Lhotse from PyTorch, i.e. make it an optional dependency. The core part of the library (except Dataset API and torchaudio augmentations) should work just fine.

Reply:
.. but in that case how will we enforce the requirement of torchaudio?  would it be part of the install instructions?

Reply:
We wouldn't enforce it, which I agree is not ideal...

I have another idea. If torch is installed, we'll detect it, and change the line in requirements to `torch==<installed-version>`, and leave torchaudio without a specified version. Then pip will have to match torchaudio to torch's version. Would that work?

Reply:
> I have another idea. If torch is installed, we'll detect it, and change the line in requirements to torch==<installed-version>, and leave torchaudio without a specified version. Then pip will have to match torchaudio to torch's version. Would that work?

I believe not. I just ran a simple test on my computer and verify that it does not work:
```
(py3) fangjuns-MacBook-Air:~ fangjun$ python3 -c "import torch; print(torch.__version__)"
1.7.0
(py3) fangjuns-MacBook-Air:~ fangjun$ pip install torchaudio
Looking in indexes: https://mirrors.aliyun.com/pypi/simple/
Collecting torchaudio
  Using cached https://mirrors.aliyun.com/pypi/packages/82/72/4ba3300f2d74c56e1c2a6ea3882dd17ef47c90a46fc65271ea4a93dd48cb/torchaudio-0.9.0-cp37-cp37m-macosx_10_9_x86_64.whl (1.5 MB)
Collecting torch==1.9.0
  Downloading https://mirrors.aliyun.com/pypi/packages/51/cd/3fa53975000c37b0a0e4f5db7dd2df45749bfbc5f04c5b99812df6794511/torch-1.9.0-cp37-none-macosx_10_9_x86_64.whl (127.9 MB)
     |██                              | 7.9 MB 1.9 MB/s eta 0:01:02^C
ERROR: Operation cancelled by user
WARNING: You are using pip version 20.2.4; however, version 21.2.1 is available.
You should consider upgrading via the '/Users/fangjun/software/py3/bin/python3.7 -m pip install --upgrade pip' command.
```

Reply:
I think torchaudio and PyTorch share the same minor version. That is, if you detect that the install pytorch is 1.7.1, you can just run
```
pip install torchaudio==0.7
```
If pytorch is 1.9.0, you can run `pip install torchaudio==0.9`.
If pytorch is 1.x.y, you can run `pip install torchaudio==0.x`

Reply:
If a user is using k2, he/she must have installed PyTorch, so I agree to remove `torch` from Lhotse's requirements.txt

Reply:
I think the right command would have been `pip install torch==1.7.1 torchaudio`. Then it seems to work:

```
~ pip install torch==1.7.1
Collecting torch==1.7.1
  Using cached torch-1.7.1-cp38-none-macosx_10_9_x86_64.whl (108.9 MB)
Collecting typing-extensions
  Using cached typing_extensions-3.10.0.0-py3-none-any.whl (26 kB)
Collecting numpy
  Downloading numpy-1.21.1-cp38-cp38-macosx_10_9_x86_64.whl (16.9 MB)
     |████████████████████████████████| 16.9 MB 702 kB/s
Installing collected packages: typing-extensions, numpy, torch
Successfully installed numpy-1.21.1 torch-1.7.1 typing-extensions-3.10.0.0
(testpt) ➜  ~ pip install torch==1.7.1 torchaudio
Requirement already satisfied: torch==1.7.1 in ./miniconda3/envs/testpt/lib/python3.8/site-packages (1.7.1)
Collecting torchaudio
  Downloading torchaudio-0.9.0-cp38-cp38-macosx_10_9_x86_64.whl (1.5 MB)
     |████████████████████████████████| 1.5 MB 3.0 MB/s
Requirement already satisfied: typing-extensions in ./miniconda3/envs/testpt/lib/python3.8/site-packages (from torch==1.7.1) (3.10.0.0)
Requirement already satisfied: numpy in ./miniconda3/envs/testpt/lib/python3.8/site-packages (from torch==1.7.1) (1.21.1)
  Downloading torchaudio-0.8.1-cp38-cp38-macosx_10_9_x86_64.whl (1.5 MB)
     |████████████████████████████████| 1.5 MB 23.3 MB/s
  Downloading torchaudio-0.8.0-cp38-cp38-macosx_10_9_x86_64.whl (1.5 MB)
     |████████████████████████████████| 1.5 MB 20.9 MB/s
  Using cached torchaudio-0.7.2-cp38-cp38-macosx_10_9_x86_64.whl (1.4 MB)
Installing collected packages: torchaudio
Successfully installed torchaudio-0.7.2
```

>  I think torchaudio and PyTorch share the same minor version. That is, if you detect that the install pytorch is 1.7.1, you can just run

What if PyToch/Torchaudio move to versions 2.x and 1.x? Or patch versions go out of sync? It feels a bit too fragile to me..

> If a user is using k2, he/she must have installed PyTorch, so I agree to remove torch from Lhotse's requirements.txt

Also Lhotse has users that don't rely on k2 so ideally we'd not make it harder for them.

Reply:
(it's not ideal because it downloads all the torchaudio versions from the newest going down to the right one, but at least it seems to work)

Reply:
A user normally installs Lhotse by
```
python setup.py install
```
I guess users of Lhotse at present are mainly from k2 (snowfall), who have already installed a specific version of PyTorch. It is not convenient to use `pip install torch==1.7.1 torchaudio` to install PyTorch again.

It would be great if Lhotse solves all the dependencies automatically without user intervention and
without re-installing the already installed PyTorch.



You may get some ideas from what ESPnet is doing to solve the installation problems of torch and torchaudio
by looking at its `setup.py`. (Note: there is no `requirements.txt` in it)

Reply:
> What if PyToch/Torchaudio move to versions 2.x and 1.x?  

It requires only minor effort to update setup.py it this happens.

> Or patch versions go out of sync? 

You can ask people from torchaudio whether it is
possible that `patch versions go out of sync` (I don't think it will happen).

Reply:
> .. but in that case how will we enforce the requirement of torchaudio? would it be part of the install instructions?

I think we can enforce it inside `setup.py`. The version of torchaudio should be determined by the version
of PyTorch. The version of PyTorch is chosen by the user (by detecting that PyTorch is already installed).

Reply:
>  I hate the idea of keeping track of which version of pytorch corresponds to which version of torchaudio. It would mean every time they release even a minor update, I'd have to update Lhotse's setup.py.

You don't have to if there is a minor update, I think, at least for the current version of torchaudio and PyTorch.

Reply:
> I guess users of Lhotse at present are mainly from k2 (snowfall), who have already installed a specific version of PyTorch. It is not convenient to use `pip install torch==1.7.1 torchaudio` to install PyTorch again.
> 
> It would be great if Lhotse solves all the dependencies automatically without user intervention and
> without re-installing the already installed PyTorch.

Precisely that’s the idea — lhotse would modify the requirements to ``torch==1.7.1 torchaudio`` internally within setup.py, so it all happens automatically. I will also look at what ESPnet does.

Reply:
https://github.com/lhotse-speech/lhotse/pull/348